wgpu-native/src/device.rs

```diff
@@ -269,34 +271,39 @@ impl DestroyedResources<back::Backend> {
     }
 
     fn handle_mapping(&mut self, raw: &<back::Backend as hal::Backend>::Device) {
-        let mut buffer_guard = HUB.buffers.write();
-
         for buffer_id in self.ready_to_map.drain(..) {
-            let buffer = &mut buffer_guard[buffer_id];
             let mut operation = None;
-            std::mem::swap(&mut operation, &mut buffer.pending_map_operation);
-            match operation {
-                Some(BufferMapOperation::Read(range, callback, userdata)) => {
-                    if let Ok(ptr) = unsafe { raw.map_memory(&buffer.memory, range.clone()) } {
-                        if !buffer.memory_properties.contains(hal::memory::Properties::COHERENT) {
-                            unsafe { raw.invalidate_mapped_memory_ranges(iter::once((&buffer.memory, range.clone()))).unwrap()  }; // TODO
+            let (result, ptr) = {
+                let mut buffer_guard = HUB.buffers.write();
+                let buffer = &mut buffer_guard[buffer_id];
+                std::mem::swap(&mut operation, &mut buffer.pending_map_operation);
+                match operation.clone().unwrap() {
```

Comment:
nit: it looks like we could combine both `match operation`s and just call the callback directly for each of the four branches. Then we could skip the `clone` here (possibly the derive too?) and the `ptr.unwrap_or`s below

Reply:
We can't invoke the callbacks directly here, because we are holding a mutable lock on `buffer_guard`. The matches are separated so that this lock is released before we invoke the callback.

Reply:
Ah, yeah that's a bit tricky. Sounds good.